packages/api-providers/src/abstracts/relayer/webb-relayer.ts

```diff
@@ -0,0 +1,262 @@
+
+    ws.onmessage = ({ data }) => {
+      const handledMessage = this.handleMessage(JSON.parse(data));
+
+      this.status = handledMessage[0];
+      this.emitter.next(handledMessage);
+
+      if (this.status === RelayedWithdrawResult.CleanExit) {
+        this.emitter.complete();
+        this.ws.close();
+      }
+    };
+
+    ws.onerror = (e) => {
+      console.log(e);
+    };
+  }
+
+  private handleMessage = (data: RelayerMessage): [RelayedWithdrawResult, string | undefined] => {
+    if (data.error || data.withdraw?.errored) {
+      return [RelayedWithdrawResult.Errored, data.error || data.withdraw?.errored?.reason];
+    } else if (data.network === 'invalidRelayerAddress') {
+      return [RelayedWithdrawResult.Errored, 'Invalid relayer address'];
+    } else if (data.withdraw?.finalized) {
+      return [RelayedWithdrawResult.CleanExit, data.withdraw.finalized.txHash];
+    } else {
+      return [RelayedWithdrawResult.Continue, undefined];
+    }
+  };
+
+  generateWithdrawRequest<T extends RelayedChainInput, C extends CMDSwitcher<T['baseOn']>> (
+    chain: T,
+    proof: RelayerCMDs<T['baseOn'], C>['proof'],
+    args: WithdrawRelayerArgs<T['baseOn'], C>
+  ) {
+    return {
+      [chain.baseOn]: {
+        [this.prefix]: {
+          contract: chain.contractAddress,
+          proof,
+          ...args
+        }
+      }
+    };
+  }
+
+  send (withdrawRequest: any) {
+    if (this.status !== RelayedWithdrawResult.PreFlight) {
+      throw Error('there is a withdraw process running');
+    }
+
+    this.status = RelayedWithdrawResult.OnFlight;
+    this.ws.send(JSON.stringify(withdrawRequest));
+  }
+
+  await () {
+    return this.watcher
+      .pipe(
+        filter(([next]) => {
+          return next === RelayedWithdrawResult.CleanExit || next === RelayedWithdrawResult.Errored;
+        })
+      )
+      .toPromise();
+  }
+
+  get currentStatus () {
+    return this.status;
+  }
+}
+
+export class WebbRelayer {
+  constructor (readonly endpoint: string, readonly capabilities: Capabilities) {}
+
+  async initWithdraw<Target extends RelayerCMDKey> (target: Target) {
+    const ws = new WebSocket(this.endpoint.replace('http', 'ws') + '/ws');
+
+    await new Promise((resolve, reject) => {
+      ws.onopen = resolve;
+      ws.onerror = reject;
+    });
+
+    /// insure the socket is open
+    /// maybe removed soon
+    for (;;) {
+      if (ws.readyState === 1) {
+        break;
+      }
+
+      await new Promise((resolve) => {
+        setTimeout(resolve, 300);
+      });
+    }
+
+    return new RelayedWithdraw(ws, target);
+  }
+
+  async getIp (): Promise<string> {
+    const req = await fetch(`${this.endpoint}/api/v1/ip`);
+
+    if (req.ok) {
+      return req.json();
+    } else {
+      throw new Error('network error');
+    }
+  }
+
+  // chainId should be formatted as a hex string
+  async getLeaves (chainId: string, contractAddress: string): Promise<RelayerLeaves> {
```

Comment:
Could we start adding validation given the comment specifies what the chainId format should be? Should it be a 0x prefixed hex string? We can do this in typescript using templates string types.

Reply:
I changed the function signature to accept number for chainId. Then call toString(16) before formatting into the url.